In [1]:
import numpy as np 
import matplotlib.pyplot as plt
import scipy 
import scipy.linalg  

In [21]:
A = np.array([[1, 2, 3], [2, 16, 21], [4, 28, 73]])
P, L, U = scipy.linalg.lu(A)

print(P)
print(L)
print(U)

print(np.dot(P.transpose(), A) - np.dot(L, U))

[[0. 1. 0.]
 [0. 0. 1.]
 [1. 0. 0.]]
[[ 1.    0.    0.  ]
 [ 0.25  1.    0.  ]
 [ 0.5  -0.4   1.  ]]
[[  4.    28.    73.  ]
 [  0.    -5.   -15.25]
 [  0.     0.   -21.6 ]]
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


In [22]:
np.linalg.matrix_rank(A)

3

In [20]:
C = np.array([[1, 2, 3], [2, 16, 21], [4, 28, 73], [0, 5, -1]])
np.linalg.matrix_rank(C)

3

Пусть $b = \left(
\begin{array}{cccc}
0\\5\\-1
\end{array}
\right)$
### $P \cdot A \cdot x = L \cdot U \cdot x = b$
### $\left(
\begin{array}{cccc}
1 & 0 & 0\\
0.25 & 1 & 0\\
0.5 & -0.4 & 1
\end{array}
\right) \cdot 
\left(
\begin{array}{cccc}
y_0\\y_1\\y_2
\end{array}
\right) = \left(
\begin{array}{cccc}
-1\\0\\5
\end{array}
\right)$
$y_0 = -1$  
$0.25y_0 + y_1 = 0, y_1 = 0.25$  
$0.5y_0 - 0.4y_1 + y_2 = 5, y_2 = 5.6$

### $\left(
\begin{array}{cccc}
4 & 28 & 73\\
0 & -5 & -15.25\\
0 & 0 & -21.6
\end{array}
\right)\cdot 
\left(
\begin{array}{cccc}
x_0\\x_1\\x_2
\end{array}
\right) = \left(
\begin{array}{cccc}
-1\\0.25\\5.6
\end{array}
\right)$
$-21.6x_2 = 5.6, x_2 = -\frac{56}{216} = -\frac{7}{27}$  
$-5x_1 - 15.25x_2 = 0.25, -5x_1 = \frac{1}{4} - \frac{61 \cdot 7}{27 \cdot 4}, x_1 = -\frac{27 - 427}{27 \cdot 20} = \frac{20}{27}$
$4x_0 + 28x_1 + 73x_2 = -1, x_0 = -\frac{27 + 28 \cdot 20 - 73 \cdot 7}{27 \cdot 4} = -\frac{76}{27 \cdot 4} = -\frac{19}{27}$

### $\left(
\begin{array}{cccc}
x_0\\x_1\\x_2
\end{array}
\right) = \left(
\begin{array}{cccc}
-\frac{19}{27}\\\frac{20}{27}\\-\frac{7}{27}
\end{array}
\right)$

In [24]:
A = np.array([[1, 2, 3], [2, 16, 21], [4, 28, 73]])
B = np.array([0, 5, -1])

np.linalg.solve(A, B)

array([-0.7037037 ,  0.74074074, -0.25925926])